File: apps/windows/TeslaSync.App/feature-views/BatteryTab/BatteryTab.Model.cs
using TeslaSync.App.Core.Charts;
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.FeatureViews.Analytics;

/// <summary>
/// One battery-health trend sample — the native mirror of a single element of the web
/// <c>FleetAnalytics.battery_trend</c> array
/// (<c>web/src/api/types.ts</c>: <c>{ date; health_score; capacity_wh; degradation_pct; range_km;
/// cycle_count }</c>). Values are SI on the wire: <see cref="CapacityWh"/> is watt-hours,
/// <see cref="RangeKm"/> is the backend's SI kilometres, <see cref="HealthScore"/> /
/// <see cref="DegradationPct"/> are percentages and <see cref="CycleCount"/> is a count. Pure data —
/// no WinUI types — so the projection is unit-tested without a UI host.
/// </summary>
public sealed record BatteryTrendPoint(
    string Date,
    double HealthScore,
    double CapacityWh,
    double DegradationPct,
    double RangeKm,
    double CycleCount);

/// <summary>
/// The render-time data model the <c>BatteryTab</c> surface binds to — the native analogue of the web
/// component's only prop (<c>data: FleetAnalytics | undefined</c>, from which it reads
/// <c>battery_trend</c>). The web component is presentational: the parent analytics page owns the query
/// lifecycle, so this model carries only the parent's <see cref="Loading"/> flag and the resolved
/// <see cref="Trend"/> series. User-facing labels and the unit preference are supplied to the projection,
/// not stored here. Pure data — no WinUI types.
/// </summary>
public sealed record BatteryTabModel(bool Loading, IReadOnlyList<BatteryTrendPoint> Trend)
{
    /// <summary>The initial model: the parent's first fetch is in flight and no trend has arrived yet.</summary>
    public static BatteryTabModel Pending { get; } = new(true, Array.Empty<BatteryTrendPoint>());

    /// <summary>A resolved model with no trend rows — the empty state.</summary>
    public static BatteryTabModel Empty { get; } = new(false, Array.Empty<BatteryTrendPoint>());
}

/// <summary>
/// The mutually-exclusive render branch of the <c>BatteryTab</c> surface — the native union of the states
/// the web component renders (<c>features/analytics/components/analytics/BatteryTab.tsx</c>). The web
/// source is presentational (it takes the analytics <c>data</c> as a prop and performs no fetching), so the
/// branches are a direct function of the input <see cref="BatteryTabModel"/>: there is no fetch-driven
/// error / stale / offline branch to reproduce here — the parent analytics page owns the query lifecycle
/// (loading spinner, <c>QueryError</c> retry, stale/offline chrome) and only mounts this surface with the
/// resolved data. Every branch maps onto a visible surface; none is ever hidden.
/// </summary>
public enum BatteryTabState
{
    /// <summary>The parent fetch is in flight (web <c>data === undefined</c>) — skeleton chrome.</summary>
    Loading,

    /// <summary>Resolved with no trend rows (web <c>trend.length === 0</c>) — the friendly empty state.</summary>
    Empty,

    /// <summary>At least one trend row (web fall-through) — the metric cards + four chart panels.</summary>
    Ready,
}

/// <summary>
/// Which native cartesian control renders a <see cref="BatteryChartPanel"/> — the direct mapping of the web
/// recharts element the panel uses: <see cref="Area"/> ↔ <c>AreaChart</c>, <see cref="Line"/> ↔
/// <c>LineChart</c>, <see cref="Composed"/> ↔ <c>ComposedChart</c> (an area + a line on one surface).
/// </summary>
public enum BatteryChartKind
{
    /// <summary>Single soft-area series (web <c>AreaChart</c>).</summary>
    Area,

    /// <summary>Single line series (web <c>LineChart</c>).</summary>
    Line,

    /// <summary>Mixed area + line series on one surface (web <c>ComposedChart</c>).</summary>
    Composed,
}

/// <summary>
/// One projected, render-ready battery metric tile — the native analogue of a single web <c>MetricCard</c>
/// in the tab's hero grid. <see cref="Value"/> is the pre-formatted headline (the web <c>value</c>),
/// <see cref="Subtitle"/> is the muted unit caption beneath it (the web <c>subtitle</c>, e.g. "%" or the
/// distance unit), <see cref="AccentBrushKey"/> is the token brush key for the accent rail (mapped from the
/// web <c>color</c>), and <see cref="AutomationName"/> is the spoken "label: value unit" name. Pure data.
/// </summary>
public sealed record BatteryMetricCard(
    string Label,
    string Value,
    string Subtitle,
    string AccentBrushKey,
    string AutomationName);

/// <summary>
/// One projected, render-ready chart panel — the native analogue of one web <c>GlassPanel</c> wrapping a
/// <c>SectionTitle</c> and a recharts chart. <see cref="Title"/> is the localized section heading,
/// <see cref="Kind"/> selects the native chart control, <see cref="Series"/> is the WinUI-free data the
/// control plots (X is the ordinal sample index, matching recharts' categorical date axis),
/// <see cref="AccessibleSummary"/> is the spoken chart description, and <see cref="AutomationName"/> is the
/// panel region's Narrator name. Pure data so every panel is asserted headlessly.
/// </summary>
public sealed record BatteryChartPanel(
    string Title,
    BatteryChartKind Kind,
    IReadOnlyList<ChartSeries> Series,
    string AccessibleSummary,
    string AutomationName);

/// <summary>
/// The fully projected, render-ready view of the surface for one input model — the native analogue of what
/// the web <c>BatteryTab</c> returns. Holds the active <see cref="State"/>, the resolved surface / empty /
/// loading labels, the projected hero <see cref="Metrics"/> and the four chart <see cref="Charts"/>, plus
/// the root <see cref="AutomationName"/>. Pure data so every branch is asserted without a UI host.
/// </summary>
public sealed record BatteryTabDisplay(
    BatteryTabState State,
    string SurfaceName,
    string EmptyMessage,
    string LoadingLabel,
    IReadOnlyList<BatteryMetricCard> Metrics,
    IReadOnlyList<BatteryChartPanel> Charts,
    string AutomationName);

/// <summary>
/// Pure projection from a <see cref="BatteryTabModel"/> (+ i18n facade + unit preference) to its
/// <see cref="BatteryTabDisplay"/> — the native port of
/// <c>features/analytics/components/analytics/BatteryTab.tsx</c>. The branch precedence mirrors the web
/// source exactly (loading → empty → ready); the hero metric values render through the shared SI display
/// boundary (<see cref="NumberFormatting"/> = the web <c>fmtNumber</c>/<c>fmtInt</c>,
/// <see cref="UnitFormatters.FormatEnergy"/> = the web <c>formatEnergy</c>, and
/// <see cref="UnitConverters.DistanceFromSi"/> = the web <c>convertDistanceFromSI</c>); and the four chart
/// series mirror the web <c>dataKey</c>s and <c>CHART_COLORS</c> indices. Every label resolves through the
/// i18n facade using the same keys the web source feeds into <c>t()</c>. No WinUI types — unit-tested
/// without a UI host.
/// </summary>
public static class BatteryTabProjection
{
    /// <summary>Token brush key for the Health Score accent rail (web <c>color="green"</c>).</summary>
    public const string AccentHealth = "TsColorSuccessBrush";

    /// <summary>Token brush key for the Capacity accent rail (web <c>color="cyan"</c>).</summary>
    public const string AccentCapacity = "TsColorAccentBrush";

    /// <summary>Token brush key for the Degradation accent rail (web <c>color="amber"</c>).</summary>
    public const string AccentDegradation = "TsColorWarningBrush";

    /// <summary>Token brush key for the Est. Range accent rail (web <c>color="purple"</c>).</summary>
    public const string AccentRange = "TsChartPowerBrush";

    /// <summary>Token brush key for the Cycles accent rail (web <c>color="cyan"</c>).</summary>
    public const string AccentCycles = "TsColorAccentBrush";

    /// <summary>Categorical palette index for the Health series (web <c>CHART_COLORS[1]</c>).</summary>
    public const int ColorHealth = 1;

    /// <summary>Categorical palette index for the Capacity series (web <c>CHART_COLORS[0]</c>).</summary>
    public const int ColorCapacity = 0;

    /// <summary>Categorical palette index for the Range series (web <c>CHART_COLORS[2]</c>).</summary>
    public const int ColorRange = 2;

    /// <summary>Categorical palette index for the Degradation series (web <c>CHART_COLORS[5]</c>).</summary>
    public const int ColorDegradation = 5;

    /// <summary>Categorical palette index for the Cycle Count series (web <c>CHART_COLORS[4]</c>).</summary>
    public const int ColorCycles = 4;

    private const string Percent = "%";
    private const double MetersPerKm = 1000.0;

    /// <summary>Project <paramref name="model"/> into a render-ready display using the i18n facade and units.</summary>
    /// <param name="model">The render-time data model (the web prop).</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    /// <param name="units">The user's unit preference applied at the display boundary (the web <c>useUnits</c>).</param>
    public static BatteryTabDisplay Project(BatteryTabModel model, ILocalizer localizer, UnitPref units)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);
        ArgumentNullException.ThrowIfNull(units);

        string surfaceName = localizer.GetString("analytics.tabs.battery", "Battery");
        string emptyMessage = localizer.GetString("analytics.battery.noData", "No battery trend data available");
        string loadingLabel = localizer.GetString("common.loading", "Loading");

        BatteryTabState state = SelectState(model);
        bool ready = state == BatteryTabState.Ready;

        IReadOnlyList<BatteryMetricCard> metrics = ready
            ? BuildMetrics(model.Trend[^1], localizer, units)
            : Array.Empty<BatteryMetricCard>();
        IReadOnlyList<BatteryChartPanel> charts = ready
            ? BuildCharts(model.Trend, localizer, units)
            : Array.Empty<BatteryChartPanel>();

        string automationName = state switch
        {
            BatteryTabState.Loading => $"{surfaceName}. {loadingLabel}",
            BatteryTabState.Empty => $"{surfaceName}. {emptyMessage}",
            _ => surfaceName,
        };

        return new BatteryTabDisplay(
            State: state,
            SurfaceName: surfaceName,
            EmptyMessage: emptyMessage,
            LoadingLabel: loadingLabel,
            Metrics: metrics,
            Charts: charts,
            AutomationName: automationName);
    }

    /// <summary>Branch precedence from the web source: loading → empty → ready.</summary>
    private static BatteryTabState SelectState(BatteryTabModel model)
    {
        if (model.Loading)
        {
            return BatteryTabState.Loading;
        }

        // Web parity: `trend.length === 0` collapses an undefined/empty trend to the empty state.
        return model.Trend.Count == 0 ? BatteryTabState.Empty : BatteryTabState.Ready;
    }

    // The five hero tiles read from the LAST trend sample (web `latest = trend[trend.length - 1]`).
    private static IReadOnlyList<BatteryMetricCard> BuildMetrics(
        BatteryTrendPoint latest,
        ILocalizer localizer,
        UnitPref units)
    {
        string distanceLabel = UnitLabels.Label(units.Distance);

        // web: fmtNumber(safe(latest.health_score), 1)
        string health = NumberFormatting.Format(Safe(latest.HealthScore), units.Locale, 1);
        // web: formatEnergy(safe(latest.capacity_wh), { precision: 1 }) — useUnits forces kWh display
        // (DEFAULT_ENERGY_PREF='kWh') regardless of the metric/imperial system, so pin energy to kWh here.
        string capacity = UnitFormatters.FormatEnergy(Safe(latest.CapacityWh), units with { Energy = EnergyUnit.Kwh }, 1);
        // web: fmtNumber(safe(latest.degradation_pct), 2)
        string degradation = NumberFormatting.Format(Safe(latest.DegradationPct), units.Locale, 2);
        // web: fmtNumber(fromKm(safe(latest.range_km)), 0) where fromKm(km) = convertDistanceFromSI(km*1000, unit)
        double rangeDisplay = UnitConverters.DistanceFromSi(Safe(latest.RangeKm) * MetersPerKm, units.Distance);
        string range = NumberFormatting.Format(rangeDisplay, units.Locale, 0);
        // web: fmtInt(safe(latest.cycle_count))
        string cycles = NumberFormatting.Format(Safe(latest.CycleCount), units.Locale, 0);

        return
        [
            Metric(localizer.GetString("analytics.battery.healthScore", "Health Score"), health, Percent, AccentHealth),
            Metric(localizer.GetString("analytics.battery.capacity", "Capacity"), capacity, string.Empty, AccentCapacity),
            Metric(localizer.GetString("analytics.battery.degradation", "Degradation"), degradation, Percent, AccentDegradation),
            Metric(localizer.GetString("analytics.battery.estRange", "Est. Range"), range, distanceLabel, AccentRange),
            Metric(localizer.GetString("analytics.battery.cycles", "Cycles"), cycles, string.Empty, AccentCycles),
        ];
    }

    private static BatteryMetricCard Metric(string label, string value, string subtitle, string accentKey)
    {
        string automation = string.IsNullOrEmpty(subtitle)
            ? $"{label}: {value}"
            : $"{label}: {value} {subtitle}";
        return new BatteryMetricCard(label, value, subtitle, accentKey, automation);
    }

    private static IReadOnlyList<BatteryChartPanel> BuildCharts(
        IReadOnlyList<BatteryTrendPoint> trend,
        ILocalizer localizer,
        UnitPref units)
    {
        string distanceLabel = UnitLabels.Label(units.Distance);

        // Health Score Timeline — web AreaChart dataKey="health_score".
        var healthSeries = new ChartSeries(
            localizer.GetString("analytics.battery.health", "Health %"),
            Points(trend, static p => Safe(p.HealthScore)))
        {
            Kind = ChartSeriesKind.Area,
            ColorIndex = ColorHealth,
            Decimals = 1,
        };

        // Capacity Trend — web LineChart dataKey="capacity_wh" (plots raw SI watt-hours, unconverted).
        var capacitySeries = new ChartSeries(
            localizer.GetString("analytics.battery.capacity", "Capacity"),
            Points(trend, static p => Safe(p.CapacityWh)))
        {
            Kind = ChartSeriesKind.Line,
            ColorIndex = ColorCapacity,
            Decimals = 0,
        };

        // Range Trend — web LineChart dataKey="range" where range = fromKm(safe(d.range_km)).
        var rangeSeries = new ChartSeries(
            $"{localizer.GetString("analytics.battery.range", "Range")} ({distanceLabel})",
            Points(trend, p => UnitConverters.DistanceFromSi(Safe(p.RangeKm) * MetersPerKm, units.Distance)))
        {
            Kind = ChartSeriesKind.Line,
            ColorIndex = ColorRange,
            Decimals = 0,
            Unit = distanceLabel,
        };

        // Degradation & Cycles — web ComposedChart: Area degradation_pct (left axis) + Line cycle_count (right).
        var degradationSeries = new ChartSeries(
            localizer.GetString("analytics.battery.degradPct", "Degradation %"),
            Points(trend, static p => Safe(p.DegradationPct)))
        {
            Kind = ChartSeriesKind.Area,
            ColorIndex = ColorDegradation,
            Decimals = 2,
        };
        var cyclesSeries = new ChartSeries(
            localizer.GetString("analytics.battery.cycleCount", "Cycle Count"),
            Points(trend, static p => Safe(p.CycleCount)))
        {
            Kind = ChartSeriesKind.Line,
            ColorIndex = ColorCycles,
            Decimals = 0,
        };

        return
        [
            Panel(localizer.GetString("analytics.battery.healthTimeline", "Health Score Timeline"), BatteryChartKind.Area, [healthSeries]),
            Panel(localizer.GetString("analytics.battery.capacityTrend", "Capacity Trend"), BatteryChartKind.Line, [capacitySeries]),
            Panel(localizer.GetString("analytics.battery.rangeTrend", "Range Trend"), BatteryChartKind.Line, [rangeSeries]),
            Panel(localizer.GetString("analytics.battery.degradationCycles", "Degradation & Cycles"), BatteryChartKind.Composed, [degradationSeries, cyclesSeries]),
        ];
    }

    private static BatteryChartPanel Panel(string title, BatteryChartKind kind, IReadOnlyList<ChartSeries> series)
    {
        string summary = ChartAccessibility.Summarize(title, series);
        return new BatteryChartPanel(title, kind, series, summary, summary);
    }

    // X is the ordinal sample index (recharts spaces a categorical date axis evenly by index); the date is
    // carried on each point's label so tooltips / the accessible summary can surface it.
    private static List<ChartPoint> Points(
        IReadOnlyList<BatteryTrendPoint> trend,
        Func<BatteryTrendPoint, double> selector)
    {
        var points = new List<ChartPoint>(trend.Count);
        for (int i = 0; i < trend.Count; i++)
        {
            points.Add(new ChartPoint(i, selector(trend[i]), trend[i].Date));
        }

        return points;
    }

    // web `safe`: coerce a non-finite (null/NaN/Infinity) value to 0 before formatting.
    private static double Safe(double value) => double.IsFinite(value) ? value : 0.0;
}

/// <summary>
/// PII-safe diagnostics for the <c>BatteryTab</c> surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never a trend value, capacity, range or
/// cycle count — so a diagnostics line can never leak battery telemetry. Thread-safe.
/// </summary>
public sealed class BatteryTabDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public BatteryTabDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=BatteryTab</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={BatteryTabRegistration.Slug}");
    }
}

/// <summary>
/// Canonical metadata for the <c>BatteryTab</c> feature surface — the native mirror of the web component at
/// <c>web/src/features/analytics/components/analytics/BatteryTab.tsx</c>.
/// </summary>
public static class BatteryTabRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event.</summary>
    public const string Slug = "BatteryTab";
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/BatteryTab/BatteryTabModel.kt
// Pure, framework-free model + projection for the Battery analytics feature view — the native analogue of
// everything the web component derives before returning JSX
// (web/src/features/analytics/components/analytics/BatteryTab.tsx). No Compose, no Android, no HTTP: every
// declaration here is unit-tested off-device in the :app:testReleaseUnitTest gate, keeping the composable a
// thin render layer.
//
// The web component is purely presentational — its parent (the Analytics page) loads the `FleetAnalytics`
// document and passes it down; BatteryTab reads `data?.battery_trend ?? []` and renders the latest-point
// metric cards plus four trend charts. This file owns the parts the web derives from that prop: the parsed
// trend rows, the five latest-point metric-card strings (web `fmtNumber`/`fmtInt`/`formatEnergy` over
// `safe(...)`), the per-chart plotted value series + x-axis labels (web `date.slice(5)`), and the accessible
// fallback table rows. Values stay SI on the wire; the SI -> display conversion happens here through the
// injected [UnitFormatter] (the web `useUnits` boundary), never by mutating the source.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/BatteryTab — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.batterytab

import io.teslasync.android.components.charts.ChartFormat
import io.teslasync.android.data.UiPhase
import io.teslasync.android.data.UiState
import io.teslasync.android.data.UnitFormatter
import io.teslasync.shared.core.diagnostics.Logger
import io.teslasync.shared.core.units.convertDistanceFromSI
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import java.util.Locale

/** Em dash shown wherever a value is unknown — the web `fmtNumber`/`'—'` empty marker. */
internal const val BATTERY_EM_DASH: String = "\u2014"

/** The percent sign appended to the Health Score / Degradation cards (web `subtitle="%"`). */
internal const val BATTERY_PERCENT: String = "%"

/** Card-value precision constants, mirroring the web `fmtNumber(..., n)` / `fmtInt` calls verbatim. */
internal const val HEALTH_DECIMALS: Int = 1
internal const val CAPACITY_DECIMALS: Int = 1
internal const val DEGRADATION_DECIMALS: Int = 2
internal const val RANGE_DECIMALS: Int = 0
internal const val CYCLE_DECIMALS: Int = 0

/** Meters per kilometre — the web `fromKm` helper multiplies SI km by 1000 before converting. */
internal const val METERS_PER_KM: Double = 1000.0

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object BatteryTabRegistration {
    /** Stable surface id. */
    const val ID: String = "battery-tab"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "BatteryTab"
}

/**
 * One battery-trend sample — the native mirror of the web `battery_trend[]` element
 * (`{ date, health_score, capacity_wh, degradation_pct, range_km, cycle_count }`). Numeric fields are
 * nullable so a sparse row never throws and the line charts can draw a gap across a missing sample (the
 * Android counterpart of Recharts `connectNulls`); the metric cards coerce missing values to `0` via
 * [safeValue], reproducing the web `safe(...)` guard. [capacityWh] and [rangeKm] are SI (watt-hours,
 * kilometres); conversion to the user's unit is the projection's job, never this type's.
 */
data class BatteryTrendPoint(
    val date: String,
    val healthScore: Double?,
    val capacityWh: Double?,
    val degradationPct: Double?,
    val rangeKm: Double?,
    val cycleCount: Double?,
)

/**
 * The already-localized microcopy the composable reads from the i18n catalog (P1/S10) — every `t(...)` key
 * the web component resolves. The lifecycle-chrome strings (empty / error / retry / offline / freshness /
 * table headers) are resolved inline at the Compose boundary, so this holder stays a thin content carrier.
 */
data class BatteryTabStrings(
    val healthScore: String,
    val capacity: String,
    val degradation: String,
    val estRange: String,
    val cycles: String,
    val healthTimeline: String,
    val capacityTrend: String,
    val rangeTrend: String,
    val degradationCycles: String,
    val healthSeries: String,
    val rangeSeries: String,
    val degradPct: String,
    val cycleCount: String,
)

/**
 * The fully projected, render-ready view of the battery trend — the native analogue of everything the web
 * component computes before returning JSX: the five latest-point metric-card values, the four chart value
 * series + shared x-axis labels, the resolved distance-unit label, and the accessible fallback table rows.
 * Pure data (no Compose types) so the projection is unit-tested without a UI host.
 */
data class BatteryTabDisplay(
    val isEmpty: Boolean,
    val healthScoreValue: String,
    val capacityValue: String,
    val degradationValue: String,
    val estRangeValue: String,
    val cyclesValue: String,
    val distanceUnit: String,
    val xLabels: List<String>,
    val healthValues: List<Double?>,
    val capacityValues: List<Double?>,
    val rangeValues: List<Double?>,
    val degradationValues: List<Double?>,
    val cycleValues: List<Double?>,
    val healthTable: List<List<String>>,
    val capacityTable: List<List<String>>,
    val rangeTable: List<List<String>>,
    val degradationCyclesTable: List<List<String>>,
)

/**
 * Coerces a possibly-missing/non-finite number to a finite value — the native mirror of the web `safe(...)`
 * chart helper (`Number.isFinite(v) ? v : 0`). Used for the metric-card values so a null field renders as
 * `0` rather than blanking the card, exactly as the web does.
 */
internal fun safeValue(value: Double?): Double = if (value != null && value.isFinite()) value else 0.0

/**
 * Drops the leading `YYYY-` from an ISO date to the web's `MM-DD` axis tick (web `tickFormatter={(v) =>
 * v.slice(5)}`). Inputs shorter than the prefix are returned unchanged, mirroring JavaScript `slice`.
 */
internal fun sliceDateLabel(date: String): String = if (date.length > 5) date.substring(5) else date

/**
 * Parses the `FleetAnalytics` document's `battery_trend` array — the native mirror of the web
 * `data?.battery_trend ?? []`. Tolerant of the raw snake_case the shared repo serves and of the
 * camelCased dual-shape the web client can emit (`camelCaseKeys`), and null-tolerant per field so a partial
 * row never throws. A non-object document, a missing array, or a JSON-null collapses to an empty list,
 * reproducing JavaScript optional chaining.
 */
object BatteryTrend {
    fun parse(analytics: JsonElement?): List<BatteryTrendPoint> {
        val obj = analytics as? JsonObject
        val array = (obj?.get(KEY_TREND_SNAKE) as? JsonArray) ?: (obj?.get(KEY_TREND_CAMEL) as? JsonArray)
        return array?.mapNotNull { element -> (element as? JsonObject)?.toPoint() } ?: emptyList()
    }

    private fun JsonObject.toPoint(): BatteryTrendPoint =
        BatteryTrendPoint(
            date = stringAt(KEY_DATE) ?: "",
            healthScore = doubleAt(KEY_HEALTH_SNAKE, KEY_HEALTH_CAMEL),
            capacityWh = doubleAt(KEY_CAPACITY_SNAKE, KEY_CAPACITY_CAMEL),
            degradationPct = doubleAt(KEY_DEGRADATION_SNAKE, KEY_DEGRADATION_CAMEL),
            rangeKm = doubleAt(KEY_RANGE_SNAKE, KEY_RANGE_CAMEL),
            cycleCount = doubleAt(KEY_CYCLE_SNAKE, KEY_CYCLE_CAMEL),
        )

    private fun JsonObject.doubleAt(
        snake: String,
        camel: String,
    ): Double? = (this[snake] as? JsonPrimitive)?.doubleOrNull ?: (this[camel] as? JsonPrimitive)?.doubleOrNull

    private fun JsonObject.stringAt(key: String): String? = (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

    private const val KEY_TREND_SNAKE = "battery_trend"
    private const val KEY_TREND_CAMEL = "batteryTrend"
    private const val KEY_DATE = "date"
    private const val KEY_HEALTH_SNAKE = "health_score"
    private const val KEY_HEALTH_CAMEL = "healthScore"
    private const val KEY_CAPACITY_SNAKE = "capacity_wh"
    private const val KEY_CAPACITY_CAMEL = "capacityWh"
    private const val KEY_DEGRADATION_SNAKE = "degradation_pct"
    private const val KEY_DEGRADATION_CAMEL = "degradationPct"
    private const val KEY_RANGE_SNAKE = "range_km"
    private const val KEY_RANGE_CAMEL = "rangeKm"
    private const val KEY_CYCLE_SNAKE = "cycle_count"
    private const val KEY_CYCLE_CAMEL = "cycleCount"
}

/**
 * The pure projection the composable renders — the native mirror of the web component's derivations.
 * Stateless and side-effect-free so it is fully covered by the off-device unit gate. The SI -> display
 * conversion uses the injected [UnitFormatter] (the web `useUnits` boundary): energy through
 * `formatEnergy`, distance through the shared `convertDistanceFromSI`, both keyed by the user's
 * [io.teslasync.shared.core.units.UnitPref].
 */
object BatteryTabProjection {
    /**
     * Projects [trend] for the live [formatter] into the render-ready [BatteryTabDisplay], reproducing the
     * web component's `latest` metric cards plus the four chart series + accessible tables. The metric
     * cards read the last sample (web `trend[trend.length - 1]`) through [safeValue]; the charts plot the
     * raw SI values (capacity) or the converted display value (range) over every sample in source order.
     */
    fun project(
        trend: List<BatteryTrendPoint>,
        formatter: UnitFormatter,
        locale: Locale = Locale.getDefault(),
    ): BatteryTabDisplay {
        val distanceUnit = formatter.prefs.distance.label
        val latest = trend.lastOrNull()
        val toRangeDisplay = { km: Double -> convertDistanceFromSI(km * METERS_PER_KM, formatter.prefs.distance) }

        return BatteryTabDisplay(
            isEmpty = trend.isEmpty(),
            healthScoreValue = cardNumber(latest?.healthScore, HEALTH_DECIMALS, locale, latest != null),
            capacityValue = cardEnergy(latest?.capacityWh, formatter, latest != null),
            degradationValue = cardNumber(latest?.degradationPct, DEGRADATION_DECIMALS, locale, latest != null),
            estRangeValue = cardRange(latest?.rangeKm, toRangeDisplay, locale, latest != null),
            cyclesValue = cardNumber(latest?.cycleCount, CYCLE_DECIMALS, locale, latest != null),
            distanceUnit = distanceUnit,
            xLabels = trend.map { sliceDateLabel(it.date) },
            healthValues = trend.map { it.healthScore },
            capacityValues = trend.map { it.capacityWh },
            rangeValues = trend.map { point -> point.rangeKm?.let(toRangeDisplay) },
            degradationValues = trend.map { it.degradationPct },
            cycleValues = trend.map { it.cycleCount },
            healthTable = trend.map { listOf(it.date, ChartFormat.number(it.healthScore, HEALTH_DECIMALS, locale)) },
            capacityTable = trend.map { listOf(it.date, formatter.energy(it.capacityWh, CAPACITY_DECIMALS)) },
            rangeTable =
                trend.map { point ->
                    listOf(point.date, ChartFormat.number(point.rangeKm?.let(toRangeDisplay), RANGE_DECIMALS, locale))
                },
            degradationCyclesTable =
                trend.map { point ->
                    listOf(
                        point.date,
                        ChartFormat.number(point.degradationPct, DEGRADATION_DECIMALS, locale),
                        ChartFormat.number(point.cycleCount, CYCLE_DECIMALS, locale),
                    )
                },
        )
    }

    /** A latest-point numeric card value (web `latest ? fmtNumber(safe(x), n) : '—'`). */
    private fun cardNumber(
        value: Double?,
        decimals: Int,
        locale: Locale,
        hasLatest: Boolean,
    ): String = if (!hasLatest) BATTERY_EM_DASH else ChartFormat.number(safeValue(value), decimals, locale)

    /** The capacity card value (web `latest ? formatEnergy(safe(capacity_wh), {precision:1}) : '—'`). */
    private fun cardEnergy(
        wattHours: Double?,
        formatter: UnitFormatter,
        hasLatest: Boolean,
    ): String = if (!hasLatest) BATTERY_EM_DASH else formatter.energy(safeValue(wattHours), CAPACITY_DECIMALS)

    /** The estimated-range card value (web `latest ? fmtNumber(fromKm(safe(range_km)), 0) : '—'`). */
    private fun cardRange(
        rangeKm: Double?,
        toRangeDisplay: (Double) -> Double,
        locale: Locale,
        hasLatest: Boolean,
    ): String =
        if (!hasLatest) {
            BATTERY_EM_DASH
        } else {
            ChartFormat.number(toRangeDisplay(safeValue(rangeKm)), RANGE_DECIMALS, locale)
        }
}

/** The mutually-exclusive surface drawn for a given [UiState] phase (web empty/content + the added chrome). */
enum class BatteryTabSurface { Loading, Error, Empty, Content }

/**
 * Maps a [UiState] onto the surface to render. Stale/offline cached data stays [BatteryTabSurface.Content]
 * (plus a freshness chip), never a blanked surface — the honest "last known" contract the sibling surfaces
 * follow.
 */
fun batteryTabSurface(state: UiState<*>): BatteryTabSurface =
    when (state.phase) {
        UiPhase.Loading -> BatteryTabSurface.Loading
        UiPhase.Error -> BatteryTabSurface.Error
        UiPhase.Empty -> BatteryTabSurface.Empty
        UiPhase.Content -> BatteryTabSurface.Content
    }

/**
 * Builds the cache-then-network [UiState] for the web-parity entry that takes the raw `FleetAnalytics`
 * document + a `loading` flag (web `data` / `isLoading` props): a first load with nothing parsed yet is
 * [UiPhase.Loading]; an empty trend is [UiPhase.Empty] (web `trend.length === 0`); otherwise the parsed
 * rows render as [UiPhase.Content].
 */
fun batteryTabStateOf(
    analytics: JsonElement?,
    loading: Boolean,
): UiState<List<BatteryTrendPoint>> {
    val trend = BatteryTrend.parse(analytics)
    val phase =
        when {
            loading && trend.isEmpty() -> UiPhase.Loading
            trend.isEmpty() -> UiPhase.Empty
            else -> UiPhase.Content
        }
    return UiState(phase = phase, data = trend)
}

/**
 * Emits the one PII-safe `view.opened` diagnostic with the surface [BatteryTabRegistration.SLUG] (P1/S11).
 * Kept free of Compose so it is unit-tested with a recording [Logger]; the composable calls it from its
 * first-composition effect.
 */
fun recordBatteryTabOpened(logger: Logger) {
    logger.info(EVENT_VIEW_OPENED, mapOf("surface" to BatteryTabRegistration.SLUG))
}

private const val EVENT_VIEW_OPENED = "view.opened"
